File: valid_parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        char_set = {"}": "{", "]": "[", ")": "("}
        stack = []

        for c in s:
            if c in char_set:
                if stack and stack[-1] == char_set[c]:
                    stack.pop()
                else:
                    return False
            else:
                stack.append(c)

        return True if not stack else False

File: validParentheses.js
class Solution {
  /**
   * @param {string} s
   * @return {boolean}
   */
  isValid(s) {
    let charMap = { "}": "{", "]": "[", ")": "(" };
    let stack = [];

    for (const c of s) {
      if (c in charMap) {
        if (stack && stack[stack.length - 1] === charMap[c]) {
          stack.pop();
        } else {
          return false;
        }
      } else {
        stack.push(c);
      }
    }

    return stack.length === 0;
  }
}
